### ISSUE TITLE
Improve understanding and visibility of Instance Failures and Autoscaling

### ISSUE THREAD

**Issue by frsechet** [convox/rack#718]
Hello,

I have set a rack to have 3 instances. However yesterday I received a capacity warning, and only two instances were available, the rack being unable to launch a new instance and keep it running. It would launch one instance, keep it for a few minutes, and terminate it - or sometimes not even try to launch new instances.

I then tried setting autoscaling via the console (see #717) or CLI, but it kept the rack count at 2 instances with still about 10-15 tasks failing to launch for lack of space.

I do not have the issue anymore now, as I unset autoscaling, and required many more instances than I need, and they seem to stick.

Thanks,
François

**Comment by nzoschke** [comment]
It might be a bit too late, but when instances are failing to come up you can access the system logs in the EC2 console and see if there is anything out of the ordinary.

One possible cause here would be something transient causing the convox UserData or any custom InstanceBootCommand scripts to fail and therefore to not stabilize in the AutoScalingGroup.

Another possible cause would be the convox monitor process thinking a new instance is unhealthy because docker isn't running or its not added to the ECS cluster membership.

Why instances are terminated is more mysterious than I'd like.

One of the simpler solutions would be to uniformly log instance failures into the Convox LogGroup so we could search and audit it after the fact. The causes I know of off the top of my head:

* EC2 internal health checks failed
* Convox userdata failed
* User InstanceBootCommand or InstanceRunCommand failed
* Docker unresponsive
* Root volume full or corrupt
* Docker volume full or corrupt
* Container Instance not registered

**Comment by frsechet** [comment]
Sadly, nothing useful showed up in the logs anywhere. ECS seemed to think everything was fine, although it had many tasks with 1 desired / 0 running. It clearly launched / killed not just one but at least 5-6 instances in a row that I know of.

**Comment by nzoschke** [comment]
Some next steps:

- [ ] Document convox API log searches that offer more information about the convox monitors
- [ ] Tweak logging in convox/agent and the API monitor process so searching is consistent

**Comment by nzoschke** [comment]
We think that ASG cause of the form `At 2016-06-08T22:16:40Z an instance was taken out of service in response to a user health-check.` is an AWS internal EC2 health check failure. We will want to poll the ASG and EC2 APIs to get more info about these failure

http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/monitoring-system-instance-status-check.html

This is different from `At 2016-06-07T16:29:28Z instance i-2ac4b1b0 was taken out of service in response to a user request.`

**Comment by nzoschke** [comment]
Found some interesting things simulating a docker failure...

```bash
# convox/agent logging stopped before it marked the instance unhealthy
$ convox logs --since=20m --filter='"i-e8ed6874"' --app convox
2016-06-09T17:59:44Z agent:0.68/i-e8ed6874 docker exec.Command args=ps try=2
2016-06-09T18:00:14Z agent:0.68/i-e8ed6874 docker exec.Command args=ps try=3
2016-06-09T18:04:32Z web:20160607215357/de6f960e1a3a ns=cluster_monitor count=3 connected='i-5194ebcb,i-5e1ae31b,i-e8ed6874' healthy='i-5194ebcb,i-5e1ae31b,i-e8ed6874' marked=''
2016-06-09T18:09:32Z web:20160607215357/de6f960e1a3a ns=cluster_monitor count=3 connected='i-5194ebcb,i-5e1ae31b,i-e8ed6874' healthy='i-5194ebcb,i-5e1ae31b,i-e8ed6874' marked=''
```

```bash
# convox/agent container is stuck and not able to log
[ec2-user@ip-10-0-1-162 ~]$ sudo strace -p 3359
Process 3359 attached
write(1, "agent:0.68/i-e8ed6874 awslogs pu"..., 159
```

**Comment by nzoschke** [comment]
When docker failure is detected properly it logs like:

> 2016-06-09T19:00:36Z agent:0.68/i-c31e9b5f docker ok=false count#DockerError err="signal: killed"

**Comment by nzoschke** [comment]
If both docker and convox/agent fail, it does look like eventually the instance is deregistered in the ECS cluster with `"status": "INACTIVE"`. Then the API monitor removes it and the ASG event looks like:

> At 2016-06-09T19:25:01Z an instance was taken out of service in response to a user health-check.

**Comment by nzoschke** [comment]
I did more experimentation with stopping docker and seeing how the ecs-agent works on its own (no convox/agent or monitor). ecs-agent logs:

```
2016-06-09T21:33:12Z [INFO] Creating poll dialer module="ws client" host="ecs-t-1.us-east-1.amazonaws.com"
2016-06-09T21:33:12Z [WARN] Error getting cpu stats module="stats" err="No data in the queue" container="&{DockerID:ccfae45b7a2e4b0afd83d84e3bf2257b6440c6ebea0c0239b0bedf6a34177465}"
2016-06-09T21:33:12Z [WARN] Error getting cpu stats module="stats" err="No data in the queue" container="&{DockerID:ec645860184e4261b4bd5194d7264d36a2ae5740e46b5d410aafb94d86045f0e}"
2016-06-09T21:33:12Z [WARN] Error getting instance metrics module="tcs client" err="No task metrics to report"
2016-06-09T21:34:15Z [INFO] Saving state! module="statemanager"
2016-06-09T21:34:15Z [INFO] Stopping container module="TaskEngine" task="convox-web:94 arn:aws:ecs:us-east-1:132866487567:task/bcaed955-cfdd-484e-81b0-e0b1ed7e1920, Status: (RUNNING->STOPPED) Containers: [web (RUNNING->STOPPED),registry (RUNNING->STOPPED),]" container="web(convox/api:20160607215357) (RUNNING->STOPPED)"
2016-06-09T21:34:15Z [INFO] Stopping container module="TaskEngine" task="convox-web:94 arn:aws:ecs:us-east-1:132866487567:task/bcaed955-cfdd-484e-81b0-e0b1ed7e1920, Status: (RUNNING->STOPPED) Containers: [web (RUNNING->STOPPED),registry (RUNNING->STOPPED),]" container="registry(convox/registry:20160607215357) (RUNNING->STOPPED)"
2016-06-09T21:35:15Z [INFO] Error transitioning container module="TaskEngine" task="convox-web:94 arn:aws:ecs:us-east-1:132866487567:task/bcaed955-cfdd-484e-81b0-e0b1ed7e1920, Status: (RUNNING->STOPPED) Containers: [web (RUNNING->STOPPED),registry (RUNNING->STOPPED),]" container="web(convox/api:20160607215357) (RUNNING->STOPPED)" state="STOPPED"
2016-06-09T21:35:15Z [INFO] Error transitioning container module="TaskEngine" task="convox-web:94 arn:aws:ecs:us-east-1:132866487567:task/bcaed955-cfdd-484e-81b0-e0b1ed7e1920, Status: (RUNNING->STOPPED) Containers: [web (RUNNING->STOPPED),registry (RUNNING->STOPPED),]" container="registry(convox/registry:20160607215357) (RUNNING->STOPPED)" state="STOPPED"
2016-06-09T21:35:15Z [INFO] Error for 'docker stop' of container; assuming it's stopped anyways module="TaskEngine" task="convox-web:94 arn:aws:ecs:us-east-1:132866487567:task/bcaed955-cfdd-484e-81b0-e0b1ed7e1920, Status: (RUNNING->STOPPED) Containers: [web (STOPPED->STOPPED),registry (RUNNING->STOPPED),]"
2016-06-09T21:35:15Z [INFO] Error for 'docker stop' of container; assuming it's stopped anyways module="TaskEngine" task="convox-web:94 arn:aws:ecs:us-east-1:132866487567:task/bcaed955-cfdd-484e-81b0-e0b1ed7e1920, Status: (RUNNING->STOPPED) Containers: [web (STOPPED->STOPPED),registry (STOPPED->STOPPED),]"
2016-06-09T21:35:15Z [INFO] Adding event module="eventhandler" change="ContainerChange: arn:aws:ecs:us-east-1:132866487567:task/bcaed955-cfdd-484e-81b0-e0b1ed7e1920 web -> STOPPED, Reason DockerTimeoutError: Could not transition to stopped; timed out after waiting 1m0s, Ports [{3000 3000 0.0.0.0 0} {4443 3001 0.0.0.0 0}], Known Sent: RUNNING"
2016-06-09T21:35:15Z [INFO] Adding event module="eventhandler" change="ContainerChange: arn:aws:ecs:us-east-1:132866487567:task/bcaed955-cfdd-484e-81b0-e0b1ed7e1920 registry -> STOPPED, Reason DockerTimeoutError: Could not transition to stopped; timed out after waiting 1m0s, Ports [{3000 3100 0.0.0.0 0} {443 3101 0.0.0.0 0}], Known Sent: RUNNING"
2016-06-09T21:35:15Z [INFO] Sending container change module="eventhandler" event="ContainerChange: arn:aws:ecs:us-east-1:132866487567:task/bcaed955-cfdd-484e-81b0-e0b1ed7e1920 web -> STOPPED, Reason DockerTimeoutError: Could not transition to stopped; timed out after waiting 1m0s, Ports [{3000 3000 0.0.0.0 0} {4443 3001 0.0.0.0 0}], Known Sent: RUNNING" change="ContainerChange: arn:aws:ecs:us-east-1:132866487567:task/bcaed955-cfdd-484e-81b0-e0b1ed7e1920 web -> STOPPED, Reason DockerTimeoutError: Could not transition to stopped; timed out after waiting 1m0s, Ports [{3000 3000 0.0.0.0 0} {4443 3001 0.0.0.0 0}], Known Sent: RUNNING"
2016-06-09T21:35:15Z [INFO] Task change event module="TaskEngine" event="{TaskArn:arn:aws:ecs:us-east-1:132866487567:task/bcaed955-cfdd-484e-81b0-e0b1ed7e1920 Status:STOPPED Reason: SentStatus:RUNNING}"
2016-06-09T21:35:15Z [INFO] Saving state! module="statemanager"
2016-06-09T21:35:15Z [INFO] Adding event module="eventhandler" change="TaskChange: arn:aws:ecs:us-east-1:132866487567:task/bcaed955-cfdd-484e-81b0-e0b1ed7e1920 -> STOPPED, Known Sent: RUNNING"
2016-06-09T21:35:15Z [INFO] Sending container change module="eventhandler" event="ContainerChange: arn:aws:ecs:us-east-1:132866487567:task/bcaed955-cfdd-484e-81b0-e0b1ed7e1920 registry -> STOPPED, Reason DockerTimeoutError: Could not transition to stopped; timed out after waiting 1m0s, Ports [{3000 3100 0.0.0.0 0} {443 3101 0.0.0.0 0}], Known Sent: RUNNING" change="ContainerChange: arn:aws:ecs:us-east-1:132866487567:task/bcaed955-cfdd-484e-81b0-e0b1ed7e1920 registry -> STOPPED, Reason DockerTimeoutError: Could not transition to stopped; timed out after waiting 1m0s, Ports [{3000 3100 0.0.0.0 0} {443 3101 0.0.0.0 0}], Known Sent: RUNNING"
2016-06-09T21:35:15Z [INFO] Sending task change module="eventhandler" event="TaskChange: arn:aws:ecs:us-east-1:132866487567:task/bcaed955-cfdd-484e-81b0-e0b1ed7e1920 -> STOPPED, Known Sent: RUNNING" change="TaskChange: arn:aws:ecs:us-east-1:132866487567:task/bcaed955-cfdd-484e-81b0-e0b1ed7e1920 -> STOPPED, Known Sent: RUNNING"
2016-06-09T21:35:25Z [INFO] Saving state! module="statemanager"
2016-06-09T21:56:12Z [INFO] Creating poll dialer module="ws client" host="ecs-t-1.us-east-1.amazonaws.com"
2016-06-09T21:56:12Z [WARN] Error getting cpu stats module="stats" err="No data in the queue" container="&{DockerID:ccfae45b7a2e4b0afd83d84e3bf2257b6440c6ebea0c0239b0bedf6a34177465}"
2016-06-09T21:56:12Z [WARN] Error getting cpu stats module="stats" err="No data in the queue" container="&{DockerID:ec645860184e4261b4bd5194d7264d36a2ae5740e46b5d410aafb94d86045f0e}"
2016-06-09T21:56:12Z [WARN] Error getting instance metrics module="tcs client" err="No task metrics to report"
```

At some point after this container instance starts reporting as not connected:

```json
{
    "failures": [], 
    "containerInstances": [
        {
            "status": "ACTIVE", 
            "registeredResources": [
                {
                    "integerValue": 2048, 
                    "longValue": 0, 
                    "type": "INTEGER", 
                    "name": "CPU", 
                    "doubleValue": 0.0
                }, 
                {
                    "integerValue": 7986, 
                    "longValue": 0, 
                    "type": "INTEGER", 
                    "name": "MEMORY", 
                    "doubleValue": 0.0
                }, 
                {
                    "name": "PORTS", 
                    "longValue": 0, 
                    "doubleValue": 0.0, 
                    "stringSetValue": [
                        "22", 
                        "2376", 
                        "2375", 
                        "51678"
                    ], 
                    "type": "STRINGSET", 
                    "integerValue": 0
                }, 
                {
                    "name": "PORTS_UDP", 
                    "longValue": 0, 
                    "doubleValue": 0.0, 
                    "stringSetValue": [], 
                    "type": "STRINGSET", 
                    "integerValue": 0
                }
            ], 
            "ec2InstanceId": "i-a9109435", 
            "agentConnected": false, 
            "containerInstanceArn": "arn:aws:ecs:us-east-1:132866487567:container-instance/d3176844-159f-459e-938d-40de2fb8a553", 
            "pendingTasksCount": 0, 
            "remainingResources": [
                {
                    "integerValue": 2048, 
                    "longValue": 0, 
                    "type": "INTEGER", 
                    "name": "CPU", 
                    "doubleValue": 0.0
                }, 
                {
                    "integerValue": 7986, 
                    "longValue": 0, 
                    "type": "INTEGER", 
                    "name": "MEMORY", 
                    "doubleValue": 0.0
                }, 
                {
                    "name": "PORTS", 
                    "longValue": 0, 
                    "doubleValue": 0.0, 
                    "stringSetValue": [
                        "22", 
                        "2376", 
                        "2375", 
                        "51678"
                    ], 
                    "type": "STRINGSET", 
                    "integerValue": 0
                }, 
                {
                    "name": "PORTS_UDP", 
                    "longValue": 0, 
                    "doubleValue": 0.0, 
                    "stringSetValue": [], 
                    "type": "STRINGSET", 
                    "integerValue": 0
                }
            ], 
            "runningTasksCount": 0, 
            "attributes": [
                {
                    "name": "com.amazonaws.ecs.capability.privileged-container"
                }, 
                {
                    "name": "com.amazonaws.ecs.capability.docker-remote-api.1.17"
                }, 
                {
                    "name": "com.amazonaws.ecs.capability.docker-remote-api.1.18"
                }, 
                {
                    "name": "com.amazonaws.ecs.capability.docker-remote-api.1.19"
                }, 
                {
                    "name": "com.amazonaws.ecs.capability.docker-remote-api.1.20"
                }, 
                {
                    "name": "com.amazonaws.ecs.capability.logging-driver.json-file"
                }, 
                {
                    "name": "com.amazonaws.ecs.capability.logging-driver.syslog"
                }, 
                {
                    "name": "com.amazonaws.ecs.capability.ecr-auth"
                }
            ], 
            "versionInfo": {
                "agentVersion": "1.8.2", 
                "agentHash": "64205ac", 
                "dockerVersion": "DockerVersion: 1.9.1"
            }
        }
    ]
}```

It's possible that we can delegate reporting bad docker completely to the ecs-agent and `"agentConnected": false`

**Comment by nzoschke** [comment]
Also today there was an EC2 / ASG outage. The ASG events look like:

> Description:DescriptionLaunching a new EC2 instance. Status Reason: EC2 Request limit exceeded. Launching EC2 instance failed.
> Cause:CauseAt 2016-06-09T20:19:09Z an instance was started in response to a difference between desired and actual capacity, increasing the capacity from 1 to 3.

**Comment by nzoschke** [comment]
This was released with https://github.com/convox/rack/releases/tag/20160613225525

Next time you wonder why instances are terminating or failing to launch, try looking at the API logs with the special `why` filter:

```
$ convox logs --app=convox --filter=why --follow=false --since=2h
```

**Comment by nzoschke** [comment]
Re-opening this, as we've seen that the convox Autoscaler isn't doing the smartest thing w.r.t. memory calculation.

**Comment by nzoschke** [comment]
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;VCF2AL1F&#x2F;279-remedy-autoscale-memory-calculation"><img src="https:&#x2F;&#x2F;github.trello.services&#x2F;images&#x2F;trello-icon.png" width="12" height="12"> Remedy Autoscale Memory calculation</a>

**Comment by nzoschke** [comment]
We're also still seeing too high of instance replacement even with the newest AMIs and Docker version.

Instance terminating is a problem in general, and doubly a problem with Racks that are running without any slack capacity.

One thing to consider is to turn off the convox/agent docker daemon check and replacement. The ECS agent does seem to monitor and report this in the DescribeContainerInstances API. It seems to be less aggressive but that could be a good thing.

**Comment by gmelika** [comment]
Just adding my 2 cents here.  I am also getting several incidents a week of: `an instance was taken out of service in response to a user health-check`.  However, a new instance is brought up and added to the rack as expected.  But the frequency of the health check failure is alarming and very doubtful to be a real failure.

**Comment by nzoschke** [comment]
Ok thanks for the reports. It's time to look at the monitoring again, as this is critical stuff.

Here's some data from a serious production rack:

```
$ convox logs --app convox --filter=why --since=96h
2016-07-15T21:59:37Z web:20160712212601/31186e381605 who="convox/monitor" what="marked instance i-a8af3738 unhealthy" why="ECS reported agent disconnected"
2016-07-15T22:04:37Z web:20160712212601/31186e381605 who="EC2/ASG" what="Launching a new EC2 instance: i-eb17707b" why="At 2016-07-15T22:00:12Z an instance was started in response to a difference between desired and actual capacity, increasing the capacity from 11 to 12."
2016-07-16T00:24:37Z web:20160712212601/31186e381605 who="convox/monitor" what="marked instance i-325ab0ad unhealthy" why="ECS reported agent disconnected"
2016-07-16T00:29:37Z web:20160712212601/31186e381605 who="EC2/ASG" what="Launching a new EC2 instance: i-01a1659e" why="At 2016-07-16T00:25:25Z an instance was started in response to a difference between desired and actual capacity, increasing the capacity from 11 to 12."
2016-07-17T18:24:37Z web:20160712212601/31186e381605 who="convox/monitor" what="marked instance i-3e6127b8 unhealthy" why="ECS reported agent disconnected"
2016-07-17T18:29:37Z web:20160712212601/31186e381605 who="EC2/ASG" what="Launching a new EC2 instance: i-28edb1ae" why="At 2016-07-17T18:25:25Z an instance was started in response to a difference between desired and actual capacity, increasing the capacity from 11 to 12."
2016-07-18T00:24:37Z web:20160712212601/31186e381605 who="convox/monitor" what="marked instance i-546422d2 unhealthy" why="ECS reported agent disconnected"
2016-07-18T00:29:37Z web:20160712212601/31186e381605 who="EC2/ASG" what="Launching a new EC2 instance: i-75da85f3" why="At 2016-07-18T00:25:25Z an instance was started in response to a difference between desired and actual capacity, increasing the capacity from 11 to 12."
2016-07-18T03:39:37Z web:20160712212601/31186e381605 who="convox/monitor" what="marked instance i-28edb1ae unhealthy" why="ECS reported agent disconnected"
2016-07-18T03:44:37Z web:20160712212601/31186e381605 who="EC2/ASG" what="Launching a new EC2 instance: i-3f4917b9" why="At 2016-07-18T03:40:22Z an instance was started in response to a difference between desired and actual capacity, increasing the capacity from 11 to 12."
2016-07-18T06:19:37Z web:20160712212601/31186e381605 who="convox/monitor" what="marked instance i-6950aef6 unhealthy" why="ECS reported agent disconnected"
2016-07-18T06:24:37Z web:20160712212601/31186e381605 who="EC2/ASG" what="Launching a new EC2 instance: i-86945b19" why="At 2016-07-18T06:20:31Z an instance was started in response to a difference between desired and actual capacity, increasing the capacity from 11 to 12."
2016-07-19T07:09:37Z web:20160712212601/31186e381605 who="convox/monitor" what="marked instance i-01a1659e unhealthy" why="ECS reported agent disconnected"
2016-07-19T07:14:37Z web:20160712212601/31186e381605 who="EC2/ASG" what="Launching a new EC2 instance: i-7b8257e4" why="At 2016-07-19T07:10:15Z an instance was started in response to a difference between desired and actual capacity, increasing the capacity from 11 to 12."```

Can y'all try this logs query on your own racks and report?

**Comment by nzoschke** [comment]
https://github.com/convox/rack/blob/master/api/workers/cluster.go

Is the monitoring code here. Please review! I'd like to get this refactored into the provider interface someday.

**Comment by frsechet** [comment]
@nzoschke for some reason I missed this notification (I was in vacation back then, might be that...).

This query doesn't return anything for me, even when I bump up the `--since` parameter.

**Comment by gmelika** [comment]
@nzoschke I must've missed this as well.  Here is the result of the query for us:

```
$ convox logs --app convox --filter=why --since=96h
2016-08-14T18:47:08Z web:20160720033913/a1ed13601010 who="convox/monitor" what="marked instance i-2fb58eb1 unhealthy" why="ECS reported agent disconnected"
2016-08-15T09:07:49Z web:20160720033913/fdf45f0dda55 who="convox/monitor" what="marked instance i-b958c041 unhealthy" why="ECS reported agent disconnected"
2016-08-15T09:12:49Z web:20160720033913/fdf45f0dda55 who="EC2/ASG" what="Launching a new EC2 instance: i-3475eecc" why="At 2016-08-15T09:08:47Z an instance was started in response to a difference between desired and actual capacity, increasing the capacity from 4 to 5."
2016-08-15T14:17:49Z web:20160720033913/fdf45f0dda55 who="convox/monitor" what="marked instance i-fdf3c86d unhealthy" why="ECS reported agent disconnected"
2016-08-15T14:22:49Z web:20160720033913/fdf45f0dda55 who="EC2/ASG" what="Launching a new EC2 instance: i-2a1300ba" why="At 2016-08-15T14:18:35Z an instance was started in response to a difference between desired and actual capacity, increasing the capacity from 4 to 5."
2016-08-16T15:57:49Z web:20160720033913/fdf45f0dda55 who="convox/monitor" what="marked instance i-3475eecc unhealthy" why="ECS reported agent disconnected"
2016-08-16T16:02:49Z web:20160720033913/fdf45f0dda55 who="EC2/ASG" what="Launching a new EC2 instance: i-671f7a9f" why="At 2016-08-16T15:58:44Z an instance was started in response to a difference between desired and actual capacity, increasing the capacity from 4 to 5."
2016-08-16T16:12:49Z web:20160720033913/fdf45f0dda55 who="convox/monitor" what="marked instance i-9be27963 unhealthy" why="ECS reported agent disconnected"
2016-08-16T16:17:49Z web:20160720033913/fdf45f0dda55 who="EC2/ASG" what="Launching a new EC2 instance: i-cad6c454" why="At 2016-08-16T16:13:23Z an instance was started in response to a difference between desired and actual capacity, increasing the capacity from 4 to 5."
2016-08-17T15:57:49Z web:20160720033913/fdf45f0dda55 who="convox/monitor" what="marked instance i-cad6c454 unhealthy" why="ECS reported agent disconnected"
2016-08-17T16:02:49Z web:20160720033913/fdf45f0dda55 who="EC2/ASG" what="Launching a new EC2 instance: i-7381e58b" why="At 2016-08-17T15:58:24Z an instance was started in response to a difference between desired and actual capacity, increasing the capacity from 4 to 5."
2016-08-18T11:57:49Z web:20160720033913/fdf45f0dda55 who="convox/monitor" what="marked instance i-b0322620 unhealthy" why="ECS reported agent disconnected"
2016-08-18T12:02:49Z web:20160720033913/fdf45f0dda55 who="EC2/ASG" what="Launching a new EC2 instance: i-674ea8f6" why="At 2016-08-18T11:58:32Z an instance was started in response to a difference between desired and actual capacity, increasing the capacity from 4 to 5."
2016-08-18T15:37:49Z web:20160720033913/fdf45f0dda55 who="convox/monitor" what="marked instance i-bd131e23 unhealthy" why="ECS reported agent disconnected"
```

**Comment by ddollar** [comment]
Should be fixed by #1173

**Comment by nzoschke** [comment]
I think this ticket got a bit confusing due to multiple ways instances get automatically managed.

I asked the ECS team about the `why="ECS reported agent disconnected"` problem. I learned that the ecs-agent expects disconnections and has its own reconnection logic. It's very possible that the convox cluster monitor is terminating the instances that are reporting disconnected before they get a chance to reconnect on their own.

I've never seen this ecs-agent behavior in the docs, perhaps we can find the behavior in the code. Then tune the convox monitor accordingly.

**Comment by nzoschke** [comment]
@beedub reported that disabling the convox cluster monitor completely has kept instances up for 3 days with no adverse effects.

That can be accomplished by manually scaling the monitor service to 0 in the ECS management console.

**Comment by beedub** [comment]
To be clear, the monitor also handles autoscaling and heartbeat. The heartbeat looks like a no-op, but disabling the monitor would also disable autoscaling as well. For us, that doesn't matter, but for others, it might

**Comment by frsechet** [comment]
I turned Autoscaling back on on a rack following @ddollar's advice and got this issue a few hours later: https://github.com/convox/rack/issues/1247. Might be linked?

**Comment by ddollar** [comment]
Tracking stability/fix in #1255
